torch/lib/THCUNN/SpatialDepthwiseConvolution.cu

```diff
@@ -164,25 +198,31 @@ __global__ void spatialDepthwiseConvolutionAccGradParameters(
   AccT grad = ScalarConvert<float, AccT>::to(0.0);
 
   // Block-stride loop over the number of elements we need to reduce
-  for (IndexType idx = threadIdx.x; idx < blockElements; idx += blockDim.x) {
+  const int laneId = threadIdx.x % WARP_SIZE;
+  const int batch = threadIdx.x / WARP_SIZE;
+  const int nwarps = blockDim.x / WARP_SIZE;
+  const int imageElements = outputWidth * outputHeight;
+  //use warp per item
+  for (int batchIdx = batch; batchIdx < batchSize; batchIdx += nwarps){  
```

Comment:
Here.

Reply:
In the original kernel, a threadblock was used to sum over NHW. I'm using a warp to sum values over HW dimension, and if batchSize is larger than the number of warps, a warp would loop over remaining batch items (e.g. if there are 8 warps, warp 0 would go over 0-8-16 etc image, warp 1 over 1-9-17 etc). Later in blockReduce all the warps will be reduced anyway, thus the full reduction will be over NHW, like it should be. That allows to get rid of one modulo operation inside the loop (because n/batchIdx now does not have to be computed through modulo, you are just looping over it), and brings in nice speed-up.